packages/send/frontend/src/lib/trpc.ts

```diff
@@ -2,53 +2,87 @@
  * This is the client-side code that uses the inferred types from the server
  */
 import logger from '@/logger';
-import { createTRPCClient, httpBatchLink, retryLink } from '@trpc/client';
+import {
+  createTRPCClient,
+  createWSClient,
+  httpBatchLink,
+  retryLink,
+  splitLink,
+  wsLink,
+} from '@trpc/client';
+
 import { AppRouter } from 'server/index';
+import { getDomainName } from './clientConfig';
+import { WS_PORT } from './config';
+
+// create persistent WebSocket connection
+
+const refreshUrl = `${import.meta.env.VITE_SEND_SERVER_URL}/api/auth/refresh`;
+const trpcUrl = `${import.meta.env.VITE_SEND_SERVER_URL}/trpc`;
+const isTesting = import.meta.env.VITE_TESTING === 'true';
+// We create a WebSocket client only if we are not in testing mode
+const wsClient = !isTesting
+  ? createWSClient({
+      url: `ws://${getDomainName(import.meta.env.VITE_SEND_SERVER_URL)}:${WS_PORT}`,
+    })
+  : null;
 
 /**
  * We only import the `AppRouter` type from the server - this is not available at runtime
  */
-
 export const trpc = createTRPCClient<AppRouter>({
   links: [
-    retryLink({
-      /**
-       * Retry strategy for failed requests:
-       * - For 401 unauthorized errors: Attempts to refresh the token and retries up to 3 times
-       * - For queries (not mutations): Retries up to 3 times
-       * - For all other cases: No retry
-       */
-      retry(opts) {
-        // Retry unauthorized requests (401) to refresh token
-        if (opts.error.data.code === 'UNAUTHORIZED') {
-          fetch(`${import.meta.env.VITE_SEND_SERVER_URL}/api/auth/refresh`, {
-            credentials: 'include',
-          })
-            .then(() => {
-              logger.info('revalidated token');
-            })
-            .catch((err) => {
-              logger.info('could not revalidate token', err);
+    splitLink({
+      condition: (op) => op.type === 'subscription',
+      // Handle non subscription requests with HTTP
+      false: [
+        retryLink({
+          /**
+           * Retry strategy for failed requests:
+           * - For 401 unauthorized errors: Attempts to refresh the token and retries up to 3 times
+           * - For queries (not mutations): Retries up to 3 times
+           * - For all other cases: No retry
+           */
+          retry(opts) {
+            // Retry unauthorized requests (401) to refresh token
+            if (opts.error.data.code === 'UNAUTHORIZED') {
+              fetch(refreshUrl, {
+                credentials: 'include',
+              })
+                .then(() => {
+                  logger.info('revalidated token');
+                })
+                .catch((err) => {
+                  logger.info('could not revalidate token', err);
+                });
+              return opts.attempts <= 3;
+            }
+            if (opts.op.type !== 'query') {
+              // Only retry queries
+              return false;
+            }
```

Comment:
Can this be at the beginning of the retry? If so you can de-dupe the `return opts.attemps <= 3` 😄